### PR TITLE
🌱 Bump Go to 1.21.5

### PR DESCRIPTION
**Pull request by chrischdi**
**What this PR does / why we need it**:

Bumps go to 1.21.5.

Upstream kubernetes did bump at https://github.com/kubernetes/kubernetes/pull/122201 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Part of #9578 

/area dependency

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,6 +1,6 @@
 module sigs.k8s.io/cluster-api
 
-go 1.20
+go 1.21
```
Comment (sbueringer) [inline go.mod:3]: WDYT about also setting toolchain like CR? k/k didn't do it.

After reading https://go.dev/doc/toolchain I'm not sure if we should or not

(https://github.com/kubernetes-sigs/controller-runtime/blob/52c391f17373684a96fa3b552d8a77b9da8ae5b5/go.mod#L5)

Comment (sbueringer) [reply go.mod:3]: > The go and toolchain lines can be thought of as specifying the version requirements for the module's dependency on the Go toolchain itself, just as the require lines in go. mod specify the version requirements for dependencies on other modules.

Based on this it sounds reasonable to set it

Comment (sbueringer) [reply go.mod:3]: Although we should maybe just set 1.21.5

Comment (chrischdi) [reply go.mod:3]: Setting toolchain to 1.21 would **force** folks who import this or newer versions of capi to use go 1.21.

Comment (sbueringer) [reply go.mod:3]: I think we already do it via `go 1.21`.

Comment (sbueringer) [reply go.mod:3]: Talked to Christian about it. We don't see a strong reason why we should set `toolchain`. We already enforce usage of Go 1.21 via `go 1.21` and we probably don't want to enforce a specific minimum go version if we don't have a strong reason to do so.

Thus I would merge as is. If someone comes up with a reason why we should set it, let us know and we follow-up

**Pull request merged**